All the Premium Stuff Except the Renames.

iHiD opened the pull request.

ErikSchierboom committed:
Pass in product when creating subscription

ErikSchierboom committed:
Pass in product when creating payment

ErikSchierboom committed:
Add premium scope

ErikSchierboom committed:
Award lifeime premium when activating insiders

ErikSchierboom committed:
Add audience for premium users

ErikSchierboom committed:
Add audience for insiders

ErikSchierboom committed:
Use proper urls for insiders notifications

ErikSchierboom committed:
Send notifications when joining premium

ErikSchierboom committed:
Check expiration

ErikSchierboom committed:
Work on updating premium

ErikSchierboom committed:
Update factory name

ErikSchierboom committed:
Add premium flair

ErikSchierboom committed:
Update lifetime eligibility

ErikSchierboom committed:
Rename user property

ErikSchierboom committed:
Premium users have all mentoring slots unlocked

ErikSchierboom committed:
Only show chat gpt for premium users

ErikSchierboom committed:
Only show dark mode for premium users

ErikSchierboom committed:
Add logic for premium updating

ErikSchierboom committed:
Add stub premium page

ErikSchierboom committed:
Use correct grace period

ErikSchierboom committed:
Update premium status when subscription is cancelled

ErikSchierboom committed:
Update premium status when subscription is created

ErikSchierboom committed:
Only update total donated for donations

ErikSchierboom committed:
Send email for becoming premium

ErikSchierboom committed:
Update premium

ErikSchierboom committed:
Test that premium_until is updated correctly

ErikSchierboom committed:
Simplify tests using traits

ErikSchierboom committed:
Fix deprecation warning

ErikSchierboom committed:
Add command to convert Stripe price to product

ErikSchierboom committed:
Determine product for Stripe payments

ErikSchierboom committed:
Detect correct product for paypal payments

ErikSchierboom committed:
Use correct product type for paypal

ErikSchierboom committed:
Simplify

ErikSchierboom committed:
Fix namespace

ErikSchierboom committed:
Infer product from subscription

ErikSchierboom committed:
Add command to create premium subscription

ErikSchierboom committed:
Move API from donations to payments

ErikSchierboom committed:
Add test for active subscription serializer

ErikSchierboom committed:
Serialize provider and product for active subscription

ErikSchierboom committed:
Update client-side urls

ErikSchierboom committed:
Allow creating premium subscriptions

ErikSchierboom committed:
Use secrets for paypal product names

ErikSchierboom committed:
Allow any 500 >= donation to count

ErikSchierboom committed:
Update premium lifetime until value

ErikSchierboom committed:
Fix tests

ErikSchierboom committed:
Add tests

ErikSchierboom committed:
Treat lifetime premium as regular donation

ErikSchierboom committed:
Update to 499 for lifetime insiders

iHiD committed:
Bump exercism config

ErikSchierboom committed:
Remove stripe lifetime

ErikSchierboom committed:
Don't defer for insiders update

iHiD committed:
Remove premium lifetime email and override User#premium? for now (#5469)

* Temporarily override premium

* Tweaks to allow lifetime insiders not premium

ErikSchierboom committed:
Add interval to subscription

ErikSchierboom committed:
Add interval to Stripe subscription

ErikSchierboom committed:
Fix Stripe sync_all tests

ErikSchierboom committed:
Support interval for GitHub subscriptions

ErikSchierboom committed:
Support interval for PayPal subscriptions

ErikSchierboom committed:
Support interval for PayPal subscriptions

iHiD committed:
Fix tests

iHiD commented on app/commands/payments/github/sponsorship/handle_created.rb:
```diff
@@ -0,0 +1,14 @@
+
+  def call
+    subscription = Payments::Github::Subscription::Create.(user, node_id, monthly_price_in_cents) unless is_one_time
+    Payments::Github::Payment::Create.(user, node_id, monthly_price_in_cents, subscription:)
```
How do you find the fact that subscription is nil despite never being defined or in scope on this line. I semi-like it and semi find it very weird in Ruby! (The code is idiomatic though!)

ErikSchierboom replied:
I'm okay with it. It looked idiomatic. I could also create a `subscription` method and be more explicit about it?

ErikSchierboom replied:
I did end up refactoring it

iHiD commented on app/commands/payments/stripe/payment/create.rb:
```diff
@@ -14,14 +14,21 @@ def initialize
+  def product
+    return subscription.product if subscription
+
+    # Premium payments are always linked to a subscription, even one-off payments
```
```suggestion
    # Premium payments are always linked to a subscription
```

As we no-longer have one-off product payments?

iHiD commented on app/commands/payments/subscription/create.rb:
```diff
@@ -2,22 +2,25 @@
     ).tap do
-      User::UpdateActiveDonationSubscription.(user)
+      User::UpdateActiveDonationSubscription.(user) if product == :donation
+      User::Premium::Update.(user) if product == :premium
```
Maybe a case on `product` or something here, just as they're mutually exclusive? I'm not fussed though.

iHiD commented on app/commands/user/insiders_status/activate.rb:
```diff
@@ -19,6 +19,7 @@ def call
 
     User::UpdateFlair.(user)
     User::Notification::CreateEmailOnly.defer(user, @notification_key)
+    User::Premium::Update.defer(user)
```
```suggestion
    User::Premium::Update.(user)
```

Move this above and undefer it as we want this happening in-band?

iHiD commented on app/commands/user/insiders_status/determine_eligibility_status.rb:
```diff
@@ -37,6 +39,13 @@ def period_reputation(period)
+  end
+
+  # TODO: update launch date with actual launch date
+  LAUNCH_DATE = Date.new(2023, 5, 1).freeze
```
```suggestion
  LAUNCH_DATE = Date.new(2023, 5, 30).freeze
```

iHiD commented on app/commands/user/insiders_status/trigger_update.rb:
```diff
@@ -11,6 +11,6 @@ def call
     end
 
-    User::InsidersStatus::Update.defer(user)
+    User::InsidersStatus::Update.(user)
```
I think this one should generally be `defer` actually. But we might just an exception for the donation route. If you revert this for now, I'll work it out when I test everything e2e properly.

iHiD commented on app/commands/user/insiders_status/update.rb:
```diff
@@ -27,6 +27,7 @@ def call
     User::SetDiscordRoles.defer(user)
     User::SetDiscourseGroups.defer(user)
     User::Notification::CreateEmailOnly.defer(user, @notification_key) if @notification_key
+    User::Premium::Update.defer(user)
```
This is the case where someone **loses** Insiders + Premium, right?

ErikSchierboom replied:
Correct

iHiD commented on app/commands/user/premium/expire.rb:
```diff
@@ -0,0 +1,7 @@
+
+  initialize_with :user
+
+  def call = User::Notification::CreateEmailOnly.defer(user, :expired_premium)
```
Do we call this anywhere yet? I'm guessing not?

ErikSchierboom replied:
We don't yet. We do now :)

iHiD commented on app/commands/user/premium/join.rb:
```diff
@@ -0,0 +1,10 @@
+    user.update!(premium_until:)
+    User::Notification::CreateEmailOnly.defer(user, :joined_premium) if FeatureFlag::PREMIUM
+  end
+end
```
This is to be called after someone completes the payment? I can't see where it's called atm (but struggling with the massive diff)

ErikSchierboom replied:
I had still to add that. My bad. Just did that.

iHiD commented on app/commands/user/premium/update.rb:
```diff
@@ -0,0 +1,48 @@
+
+  memoize
+  def last_payment_premium_until
+    return nil if last_payment.nil?
```
```suggestion
```

The line below covers this case.

iHiD commented on app/commands/user/premium/update.rb:
```diff
@@ -0,0 +1,48 @@
+    return 45.days if last_payment.subscription.overdue?
+
+    30.days
+  end
```
We need to pivot here on monthly/annual. (Either 30 days or 365 days (+15))

ErikSchierboom committed:
Store PayPal intervals

ErikSchierboom committed:
Add interval to active subscriptipn

ErikSchierboom commented on app/commands/payments/payment/create.rb:
```diff
@@ -0,0 +1,33 @@
+
+      # TODO: the donation guard means that a premium user doesn't get the supporter badge
+      # We probably do want that, right?
+      User::RegisterAsDonor.(user, Time.current) if product == :donation
```
@iHiD Should premium users get the supporter badge?

iHiD replied:
Not for their premium subscription payments.

ErikSchierboom committed:
Minor refactoring

ErikSchierboom committed:
Update app/commands/payments/stripe/payment/create.rb

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

ErikSchierboom committed:
Another minor refactoring

ErikSchierboom committed:
Update launch date

ErikSchierboom committed:
Use defer premium update when triggering insiders update

ErikSchierboom committed:
Use premium commands to expire and join:

ErikSchierboom committed:
Small simplification

ErikSchierboom committed:
Use interval for premium until

ErikSchierboom committed:
Allow accessing payments from a subscription

ErikSchierboom committed:
Add tests for subscription

ErikSchierboom committed:
Add method to get last payment of subscription

ErikSchierboom committed:
Update active donation subscription only considers donations

ErikSchierboom committed:
Add initial premium settings

ErikSchierboom committed:
Show payment provider used

ErikSchierboom committed:
Conditionally show receipt URL

ErikSchierboom committed:
Show last payment on top

ErikSchierboom committed:
Only show update amount and cancel link for Stripe

ErikSchierboom committed:
Add active premium subscription

ErikSchierboom committed:
Only count donations for total_donated_in_dollars

ErikSchierboom committed:
More work for premium settings

ErikSchierboom committed:
Allow parameterizing of active subscription

ErikSchierboom committed:
Allow parameterizing of active subscription

dem4ron committed:
Add subscription component, TODO: edit copy

ErikSchierboom committed:
Use question mark for current_active_subscription calls

ErikSchierboom committed:
Fix naming conflict

ErikSchierboom committed:
Fix tests

ErikSchierboom committed:
Fix more tests

ErikSchierboom committed:
Fix test

ErikSchierboom committed:
Bump config

ErikSchierboom committed:
Move some business logic to subscription

ErikSchierboom committed:
Guard for update_amount to only work on donations

ErikSchierboom committed:
Add basic UpdatePlan command

ErikSchierboom committed:
Refactor update plan

ErikSchierboom committed:
Add premium to users API

ErikSchierboom committed:
Use correct redirect link after canceling

iHiD commented on app/models/user_track/mentoring_slots.rb:
```diff
@@ -7,6 +7,8 @@ def has_available_mentoring_slot? = num_available_mentoring_slots.positive?
 
   memoize
   def num_locked_mentoring_slots
+    return 0 if user.premium?
```
Nice.

iHiD commented on db/migrate/20230516122816_add_premium_until_to_user_data.rb:
```diff
@@ -0,0 +1,8 @@
+    return if Rails.env.production?
+
+    add_column :user_data, :premium_until, :datetime, null: true
+    add_index :user_data, [:premium_until], if_not_exists: true
```
```suggestion
    add_index :user_data, :premium_until, if_not_exists: true
```

ErikSchierboom committed:
Remove payment_ prefix

ErikSchierboom committed:
Recheck premium for any expired

ErikSchierboom committed:
Add tests for updating stripe amount

ErikSchierboom committed:
Update plan

iHiD committed:
Premium Front End (#5468)

* Add premium svg

* Premium price cards (#5464)

* Add react component

* Add stripe element

* Add modal

* Style modal

* Fix styles

* Add success modal

* Add paypal links

* Remove unused code, add TODO

* Fix JS tests

* Tweak ChatGPT button behaviour (#5461)

* Split buttons logic into seperate components

* Tweak copy

* Add new site nav (#5457)

* Couple experimentations

* More experimentation

* Moore experimenting

* Handle focus

* Refactor

* Add tiny animation

* Handle overflows

* export focus handler

* Add toggle button

* Sync theme switches

* Add tooltip over theme-toggle button

* Add link on tooltip - WIP

* Add color to link

* Fix accessibility issues

* Experiment with keyboard focus - WIP

* Fix keyboard focus accessibility

* Handle no-view menus

* Use kwarg for submenu too

* Autofocus first element if item is not a link

* Disable firstLink focusing for now

* Rename content to view to match other places, add margin to toggle button

* Refactor theme observer

* Make observer slightly more efficient

* Avoid calling matchMedia too frequently

* Git refactor our some complexity

* Refactor each nav view into its own template

* Fix bug with clipping

* Fix issue with glitchy hovers

* Fix menu issue

* Slightly refactor hover pseudo fields

* Menu enhancements, WIP adding only view for premium

* Tweak css for premium

---------

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

* Make lifetime subscription a donation (#5474)

* Turn lifetime intent type into a one off `payment`

* Remove premium_lifetime related bits

* WIP

* Force dark theme on insiders + premium (#5477)

* Site header tweaks (#5480)

* Delete bunch of css

* Minor visual tweaks

* Hidden overflow on premium

* Stripe success flow (#5475)

* Add redirect link, remove success modal

* Copy paste welcome modal

* Tweaks

* Add insiders modal

* Use correct links

* Remove margin-top from site_footer on premium page

* Rename modals

* Clean up site-header

* Move some business logic to subscription

* Guard for update_amount to only work on donations

* Add basic UpdatePlan command

* Refactor update plan

---------

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>
Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

* Add route + view for paypal pending

* Add paypal status component

* Accessibility themes (#5473)

* Add sepia

* Forced colors

* Remove forced colors bits

* Tweak sepia name

* Process paypal payment (#5484)

* Process paypal payment

* Add redirectTo to its index file

* PremiumSubscription settings tweaks (#5482)

* Remove update from PremiumSubscription for now

* Work on subscription/donation settings - WIP

* Edit subscription forms

* Tweak appearence, refactor, import modal

* Add stripe modal details

* Premium - iHiD Tweaks (#5486)

* Fix test name

* Fix stuff

* WIP top menu

* Fix two test issues

* Work on the Premium payment modal

* Keep chipping away

* Fix locked bg color

* Update mentoring article

Update margin

Add conditional

* Tweak layout of mentoring slots

* Lots of little tweaks

* Add videos page

* Add video pages

* Final menu tweaks

* Revert all schema changes

* Try and fix schema

* Revert temporary change

* Fix typo

* Fix failing mentoring slot tests

* Fix nil error

* Minor naming refactoring

* Fix canceling of subscription having grace period

* Reduce from 5 to 3 tracks in nav

* Add donate modal on Insiders (#5493)

* Change event you listen to (#5491)

* WIP Nonsense (#5494)

* WIP Nonsense

* More nonsense that Erik has to deal with

* More stuff

* More stuff

* Fix tests

* Fix more tests

---------

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

---------

Co-authored-by: dem4ron <demaaron88@gmail.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>
Co-authored-by: Aron Demeter <66035744+dem4ron@users.noreply.github.com>

---------

Co-authored-by: dem4ron <demaaron88@gmail.com>
Co-authored-by: Aron Demeter <66035744+dem4ron@users.noreply.github.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

iHiD commented on app/commands/payments/payment/update_amount.rb:
```diff
@@ -8,6 +8,7 @@ def call
     payment.update!(amount_in_cents:)
-    User::UpdateTotalDonatedInCents.(payment.user)
+    User::UpdateTotalDonatedInCents.(payment.user) if payment.donation?
+    User::Premium::Update.(user) if payment.premium?
```
```suggestion
    if payment.donation?
      User::UpdateTotalDonatedInCents.(payment.user) 
    elsif payment.premium?
      User::Premium::Update.(user) 
    end
```

I just find this a bit clearer.

iHiD commented on app/commands/payments/paypal/interval.rb:
```diff
@@ -0,0 +1,14 @@
+      raise UnsupportedPaypalIntervalError
+    end
+  end
+end
```
I'd rather this either: 
1. Followed our command naming. So module `Payments::Paypal::Interval::CycleToString.(...)`
2. Was a model in app/models/payments/paypal/interval. Probably one that was initialize_with :payment_cycle then `from_payment_cycle` was `to_s`. Or something.

This doesn't block merging to be clear, but I think it would be good to tidy/refactor to use an existing pattern.

iHiD commented on app/commands/payments/paypal/payment/handle_web_accept.rb:
```diff
@@ -42,4 +44,5 @@ def external_id = payload["txn_id"]
   def payment_status = payload["payment_status"]
   def payer_id = payload["payer_id"]
   def payer_email = payload["payer_email"]
+  def product = :donation
```
Do we know this is a donation because `web_accept` is always a donation?

ErikSchierboom replied:
Yep. Otherwise it is a `handle_recurring_payment`

iHiD commented on app/commands/payments/paypal/product.rb:
```diff
@@ -0,0 +1,14 @@
+      raise UnknownPaypalProductError
+    end
+  end
+end
```
Same comment as on Interval.

iHiD replied:
The other option, which maybe nicer/simpler, would be to have these as `self.product_from_name` on `Payments::Paypal` instead.

ErikSchierboom replied:
I've refactored things a bit

iHiD commented on app/commands/payments/stripe/price.rb:
```diff
@@ -0,0 +1,34 @@
+
+  MONTH_AMOUNT_IN_CENTS = 999
+  YEAR_AMOUNT_IN_CENTS = 9999
+  LIFETIME_AMOUNT_IN_CENTS = 49_900
```
```suggestion
  MONTH_AMOUNT_IN_CENTS = 9_99
  YEAR_AMOUNT_IN_CENTS = 99_99
  LIFETIME_AMOUNT_IN_CENTS = 499_00
```

If rubocop tried to "fix" this, tell it to fuck off

iHiD commented on app/commands/payments/stripe/subscription/create.rb:
```diff
@@ -3,13 +3,28 @@
-  def amount_in_cents = stripe_data.items.data[0].price.unit_amount
+  def price = stripe_data.items.data[0].price
+  def amount_in_cents = price.unit_amount
+  def product = Payments::Stripe::Product.from_product_id(price.product)
```
```suggestion
  def amount_in_cents = price.unit_amount
  def product = Payments::Stripe::Product.from_product_id(price.product)
  def price = stripe_data.items.data[0].price
```

iHiD commented on app/commands/payments/subscription/update_status.rb:
```diff
@@ -0,0 +1,11 @@
+  def call
+    subscription.update!(status:)
+    User::UpdateActiveDonationSubscription.(subscription.user) if subscription.donation?
+    User::Premium::Update.(subscription.user) if subscription.premium?
```
Switch to if/elseif here?

iHiD commented on app/commands/user/insiders_status/determine_eligibility_status.rb:
```diff
@@ -2,7 +2,8 @@ class User::InsidersStatus::DetermineEligibilityStatus
   LIFETIME_REPUTATION_THRESHOLD = 1_000
-  LIFETIME_DONATIONS_THRESHOLD = 500_00
+  LIFETIME_DONATIONS_THRESHOLD = 499_00
+  LIFETIME_PREMIUM_AMOUNT = 499_00
```
This should be unified with the constant in `Payments::Stripe::Price` to be just one constant.

ErikSchierboom replied:
We should unify, but this is not stripe specific though (it also holds for paypal donations). Any ideas for a better place?

iHiD commented on app/commands/user/insiders_status/determine_eligibility_status.rb:
```diff
@@ -37,6 +39,12 @@ def period_reputation(period)
+    user.payments.where('amount_in_cents >= ?', LIFETIME_PREMIUM_AMOUNT).exists?
+  end
+
+  LAUNCH_DATE = Date.new(2023, 5, 30).freeze
```
```suggestion
  LAUNCH_DATE = Date.new(2023, 5, 31).freeze
```

iHiD committed:
Update app/commands/user/insiders_status/determine_eligibility_status.rb

iHiD commented on app/commands/user/insiders_status/update.rb:
```diff
@@ -24,14 +24,20 @@ def call
+    # These things are being called for when someone
+    # LOSES Insiders access, or when they change from normal,
+    # to lifetime. This code is not called when they GAIN Insiders.
+    # Look at Activate.() for that instead.
```
```suggestion
    # These things are being called for when someone:
    # - LOSES Insiders access; or
    # - UPGRADES from normal to lifetime
    # 
    # This code is not called when they GAIN Insiders.
    # Look at Activate.() for that instead.
```

iHiD commented on app/commands/user/premium/join.rb:
```diff
@@ -0,0 +1,18 @@
+    return unless existing_theme.nil? || existing_theme == 'light'
+
+    user.preferences.update!(theme: :dark)
+  end
```
This needs a test adding.

ErikSchierboom replied:
I guess we also need to do the inverse when expiring?

iHiD replied:
Yes, true.

iHiD commented on app/commands/user/premium/update.rb:
```diff
@@ -0,0 +1,43 @@
+
+    if expire?
+      User::Premium::Expire.(user)
+    else
```
Should this be:
```suggestion
    elsif user.premium_until.nil?
```

ErikSchierboom replied:
No, because of the "sliding window" aspect of it. If a new payment was made, the `premium_until` time needs to be updated. And the preceding `return if user.premium_until == premium_until` prevents unnecessary updating.

iHiD replied:
But I don't understand why we call `Join` a second time in that case. Shouldn't we be calling something to update `premium_until` instead?

ErikSchierboom replied:
That's basically what `Join` does

iHiD commented on app/commands/user/premium/update.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  private
+  def premium_until
```
```suggestion
  memoize
  def premium_until
```

iHiD commented on app/helpers/react_components/premium/paypal_status.rb:
```diff
@@ -0,0 +1,4 @@
+class ReactComponents::Premium::PaypalStatus < ReactComponents::ReactComponent
+  def to_s = super("premium-paypal-status", { endpoint: Exercism::Routes.api_user_path,
+                                              premium_redirect_path: Exercism::Routes.premium_path })
+end
```
I'd rather this was a normal method, not an inlined one.

iHiD commented on app/helpers/view_components/handle_with_flair.rb:
```diff
@@ -59,13 +59,15 @@ def ml_in_px
 
     TITLES = {
+      premium: 'Premium User',
       insider: 'An Insider',
```
```suggestion
      insider: 'An Exercism Insider',
```

iHiD commented on app/models/payments/payment.rb:
```diff
@@ -15,7 +15,7 @@ def email_communication_preferences_key; end
   def amount_in_dollars = amount_in_cents / BigDecimal(100)
   def self.total_supporters = select(:user_id).distinct.count
-  def self.total_donated_in_dollars = sum(:amount_in_cents) / BigDecimal(100)
+  def self.total_donated_in_dollars = donation.sum(:amount_in_cents) / BigDecimal(100)
```
This looks wrong to me (singular.sum)

iHiD commented on app/models/user.rb:
```diff
@@ -131,6 +132,8 @@ class User < ApplicationRecord
 
   scope :with_data, -> { joins(:data) }
   scope :donor, -> { with_data.where.not(user_data: { first_donated_at: nil }) }
```
This should be plural.

iHiD commented on app/models/user.rb:
```diff
@@ -131,6 +132,8 @@ class User < ApplicationRecord
 
   scope :with_data, -> { joins(:data) }
   scope :donor, -> { with_data.where.not(user_data: { first_donated_at: nil }) }
+  scope :premium, -> { with_data.where('user_data.premium_until > ?', Time.current) }
+  scope :insiders, -> { with_data.where(user_data: { insiders_status: %i[active active_lifetime] }) }
```
We need indexes on both of these columns.

ErikSchierboom replied:
They were already there:

https://github.com/exercism/website/blob/main/db/migrate/20230412075814_add_insiders_status_to_users.rb#L6
https://github.com/exercism/website/pull/5467/files#diff-b12c7e45b80830692065f74867506a622dbe92f05ac9099de2ff63015d6808b6R6

iHiD committed:
Reorder self methods to the top

iHiD committed:
Update db/migrate/20230516122816_add_premium_until_to_user_data.rb

iHiD committed:
Update app/commands/user/insiders_status/update.rb

iHiD committed:
Remove unused files

iHiD committed:
Sort aron life (#5497)

* Always show theme preference settings

* Don't update if same, but update if has been updated, copy tweaks

* Remove tr ws

---------

Co-authored-by: dem4ron <demaaron88@gmail.com>

dem4ron committed:
Payment Redirect tweaks (#5498)

* Add premium svg

* Premium price cards (#5464)

* Add react component

* Add stripe element

* Add modal

* Style modal

* Fix styles

* Add success modal

* Add paypal links

* Remove unused code, add TODO

* Fix JS tests

* Tweak ChatGPT button behaviour (#5461)

* Split buttons logic into seperate components

* Tweak copy

* Add new site nav (#5457)

* Couple experimentations

* More experimentation

* Moore experimenting

* Handle focus

* Refactor

* Add tiny animation

* Handle overflows

* export focus handler

* Add toggle button

* Sync theme switches

* Add tooltip over theme-toggle button

* Add link on tooltip - WIP

* Add color to link

* Fix accessibility issues

* Experiment with keyboard focus - WIP

* Fix keyboard focus accessibility

* Handle no-view menus

* Use kwarg for submenu too

* Autofocus first element if item is not a link

* Disable firstLink focusing for now

* Rename content to view to match other places, add margin to toggle button

* Refactor theme observer

* Make observer slightly more efficient

* Avoid calling matchMedia too frequently

* Git refactor our some complexity

* Refactor each nav view into its own template

* Fix bug with clipping

* Fix issue with glitchy hovers

* Fix menu issue

* Slightly refactor hover pseudo fields

* Menu enhancements, WIP adding only view for premium

* Tweak css for premium

---------

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

* Make lifetime subscription a donation (#5474)

* Turn lifetime intent type into a one off `payment`

* Remove premium_lifetime related bits

* WIP

* Force dark theme on insiders + premium (#5477)

* Site header tweaks (#5480)

* Delete bunch of css

* Minor visual tweaks

* Hidden overflow on premium

* Stripe success flow (#5475)

* Add redirect link, remove success modal

* Copy paste welcome modal

* Tweaks

* Add insiders modal

* Use correct links

* Remove margin-top from site_footer on premium page

* Rename modals

* Clean up site-header

* Move some business logic to subscription

* Guard for update_amount to only work on donations

* Add basic UpdatePlan command

* Refactor update plan

---------

Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>
Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

* Add route + view for paypal pending

* Add paypal status component

* Accessibility themes (#5473)

* Add sepia

* Forced colors

* Remove forced colors bits

* Tweak sepia name

* Process paypal payment (#5484)

* Process paypal payment

* Add redirectTo to its index file

* PremiumSubscription settings tweaks (#5482)

* Remove update from PremiumSubscription for now

* Work on subscription/donation settings - WIP

* Edit subscription forms

* Tweak appearence, refactor, import modal

* Add stripe modal details

* Fix test name

* Fix stuff

* WIP top menu

* Fix two test issues

* Work on the Premium payment modal

* Keep chipping away

* Fix locked bg color

* Update mentoring article

Update margin

Add conditional

* Tweak layout of mentoring slots

* Lots of little tweaks

* Add videos page

* Add video pages

* Final menu tweaks

* Revert all schema changes

* Try and fix schema

* Revert temporary change

* Fix typo

* Fix failing mentoring slot tests

* Fix nil error

* Minor naming refactoring

* Fix canceling of subscription having grace period

* Reduce from 5 to 3 tracks in nav

* Add donate modal on Insiders (#5493)

* Change event you listen to (#5491)

* Add insiders redirect link on lifetime price option

* Add await

* Remove empty lines

* Readd one line

---------

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>
Co-authored-by: Erik Schierboom <erik_schierboom@hotmail.com>

dem4ron committed:
Add Journey view (#5499)

* Add journey page starting point

* Add padding to container

iHiD committed:
Improve insiders eligibility algorithm

iHiD committed:
Small tweaks on joining flow

iHiD committed:
Add ChatGPT upsell

iHiD committed:
Add mails for insiders

iHiD committed:
Refactor and add tests

iHiD committed:
Fix more tests

ErikSchierboom committed:
Fix tests

ErikSchierboom committed:
Fix copy

ErikSchierboom committed:
Update flair when premium changes

dem4ron committed:
Pass down missing value (#5502)

ErikSchierboom committed:
Use separators for price amounts

ErikSchierboom committed:
Lots of refactorings

ErikSchierboom committed:
Update app/commands/payments/stripe/subscription/create.rb

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

ErikSchierboom committed:
Some refactoring

dem4ron committed:
Add zero tracks section view (#5501)

* Add zero tracks section view

* Dry things up a bit

dem4ron committed:
Add theme preference settings tests (#5503)

* Add theme preference settings

* Format tests, add one more case

ErikSchierboom committed:
Add tests for changing theme

ErikSchierboom committed:
Fix failing test

iHiD commented on app/commands/payments/subscription/create.rb:
```diff
@@ -22,6 +22,7 @@ def call
         User::UpdateActiveDonationSubscription.(user)
       when :premium
         User::Premium::Update.(user)
+        User::InsidersStatus::Update.(user)
```
I don't understand this line. Creating a Premium subscription shouldn't affect someone's insider status should it?

ErikSchierboom replied:
No that line can be removed. It was a test for the flair bit

ErikSchierboom committed:
Remove unused line

ErikSchierboom committed:
Separate premium update from premium join

iHiD committed:
Add extra mailers

iHiD committed:
Add extra community preferences

iHiD committed:
Set correct notification keys

ErikSchierboom committed:
Reset chatgpt usage for premium users

ErikSchierboom committed:
Schedule chatgpt usage resetting

iHiD committed:
Add Paypal waiting page

iHiD committed:
Refine upsell points

iHiD committed:
Rename route

iHiD committed:
Fix test path breakage

iHiD committed:
Fix test

dem4ron committed:
Dark theme tweaks (#5504)

* About nav colors

* Donate page, concept heading, fix wrong method name

* Automation tweaks

* Contributor credit block, contributor tags

iHiD committed:
Fix typos
